design_notes/XDN11 - Error handling.md

```diff
@@ -0,0 +1,345 @@
+Errors may also end up traveling through several different components written in different languages
+before becoming a failure. To assist with diagnosing these failures, xlang will allow a projection to
+associate additional diagnostic information to the existing information as an error travels through it.
+This will help a developer get a complete picture of the failure which wouldn't be available with just
+the information collected from the language that initially returned the error. An example of this is
+the stack trace. A language will typically only be aware of the stack within its language boundary and
+thereby the diagnostic information collected at the initial error would only have a partial stack.
+The following additional information can be associated with the error as it travels through projections:
+
+- Stack trace
+- Projection identifier
+- Any additional diagnostic information
+
+The diagnostic information will be stored in a COM object (error info) implementing the
+IXlangErrorInfo interface. As with any COM object, it will be reference counted and will implement
+IUnknown. This will allow it to be easily accessed by projections written in different languages
+while still being able to ensure the cleanup of it.
+
+## Handling language errors
+
+Producing projections may encounter a language error when projecting an API call to a component.
+The error will either be from the translation of the call to the component's language or be from
+the component itself as a result of the call. Either way, upon encountering an error, the projection
+should determine whether the error is new or is a propagation of a previous xlang error. Strategies
+on doing this will be discussed in a later section.
+
+If the error is new, the projection should translate the error to an equivalent xlang error or to
+xlang_fail if there is no equivalent mapping. After translating the error, the projection should
+call xlang's error origination API, xlang_originate_error, to store the diagnostic information
+collected by the language for the error. The API will create an error info with the provided
+diagnostic information and store it on a thread level storage. The created error info will remain
+on the thread level storage until it is retrieved by the consuming projection.
+
+xlang_originate_error will be defined in the PAL and will take the following parameters:
+- a xlang_result with the translated xlang error for the language specific error that occurred
+- a xlang_string with the language specific error that occurred
+- a xlang_string with the error message
+- a xlang_string with the error stack if available
+- a xlang_string with an identifier for the projection
+- an IUnknown pointer with any additional language specific information (can be null)
+
+```cpp
+void xlang_originate_error(
+	xlang_result error,
+	xlang_string languageError,
+	xlang_string message,
+	xlang_string stack,
+	xlang_string projectionIdentifier,
+	IUnknown* languageInformation)
```

Comment:
Should indicate that most of these are optional (and thus order them accordingly). In many cases you'll only have an error code and an error message.

Reply:
Addressed by calling out only the error is required.  But, it is still encouraged for projections to provide the other fields as it helps a developer in diagnosing the issue.